Migrate to MVP: Entry screen

VSemikoz opened the pull request:
Refactor to implement MVP arch on MainScreen

VSemikoz committed:
add BaseView and BasePresenter interfaces

VSemikoz committed:
add NewsFeedContract

VSemikoz committed:
add NewsFeedActivity and register it into manifest

VSemikoz committed:
add NewsFeedPresenter and NewsFeedFragment

VSemikoz committed:
add newsfeed_act layout

VSemikoz committed:
add carcass of NewsFeedActivity

VSemikoz committed:
add carcass of NewsFeedFragment

VSemikoz committed:
add ActivityUtils and addFragmentToActivity in it

VSemikoz committed:
add methods in contract

VSemikoz committed:
add activity functionality

VSemikoz committed:
add methods fragment and presenter implementations

VSemikoz committed:
delete MainActivity

VSemikoz committed:
delete MainActivity, MainFragment, main_activity_layout

VSemikoz committed:
add resource string tv_category_string_prefix

VSemikoz committed:
rename news_feed_fragment -> newsfeed_frag

VSemikoz committed:
change tools:context=".MainActivity" -> ".newsfeed.NewsFeedActivity"

VSemikoz committed:
add methods into contract

VSemikoz committed:
add onCategorySelected() override method

VSemikoz committed:
contract methods implementation

VSemikoz committed:
remove unnecessary comments

VSemikoz committed:
extract dialog manipulation in fragment

VSemikoz committed:
add static Navigator to extract move between screen logic

VSemikoz committed:
correct open news in browser

VSemikoz committed:
rename newsFeedView -> view

VSemikoz committed:
remove save and restore state from activity

VSemikoz committed:
update set of contracts's methods

VSemikoz committed:
contract's methods implementation

VSemikoz committed:
bug fix with news item progress bar

VSemikoz committed:
code refactor

VSemikoz committed:
correct access modifiers

VSemikoz committed:
methods rename

VSemikoz committed:
add view methods and implementations

VSemikoz committed:
change logic of get favorite and category news, progress bar implementation

VSemikoz committed:
rename method fillFragmentByView -> updateNewsListUI

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/dialogs/PickCategoryDialog.java:
```diff
@@ -33,13 +33,13 @@ public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
+
         try {
-            nListener = (OnCategorySelectedListener) getActivity();
+            this.nListener = (OnCategorySelectedListener) childFragment;
```
Why is a listener called `nListener`? Let's pick up a more descriptive name or leave it a `listener`. BTW, why is this logic still resides in the `onAttach` method? Can't we create a setter for our listener?

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/navigator/Navigator.java:
```diff
@@ -0,0 +1,34 @@
+import ru.vssemikoz.newsfeed.dialogs.PickCategoryDialog;
+
+public class Navigator {
+    static public void openWebView(String url, Context context) {
```
1. Modifiers are usually going in another sequence: {visibility mod} {static} but not {static} {visibility mod}.
2. Your methods shouldn't be static as we'll implement a DI soon and we'll construct a `Navigator` instance.
3. A `Navigator` shouldn't show dialogs. It's used to navigate the user.

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedActivity.java:
```diff
@@ -0,0 +1,31 @@
+
+        NewsFeedFragment newsFeedFragment = (NewsFeedFragment) getSupportFragmentManager().findFragmentById(R.id.newsfeed_content_frame);
+        if (newsFeedFragment == null) {
+
```
This is an empty line. Try to review your code before commits.

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedActivity.java:
```diff
@@ -0,0 +1,31 @@
+import ru.vssemikoz.newsfeed.R;
+import ru.vssemikoz.newsfeed.utils.TypeConverters.ActivityUtils;
+
+public class NewsFeedActivity extends AppCompatActivity {
```
There's no need in any logic in this activity since we can initialize a presenter in the Fragment constructor or in one of its callbacks.

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedFragment.java:
```diff
@@ -0,0 +1,216 @@
+    public NewsFeedFragment() {
+    }
+
+    static NewsFeedFragment newInstance() {
```
There's no need in creating `newInstance()` method since each `Fragment` already has it.

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedFragment.java:
```diff
@@ -0,0 +1,216 @@
+
+    @Override
+    public void onCreate(@Nullable Bundle savedInstanceState) {
+        Log.d(TAG, "onCreate: ");
```
Do we need this logging?

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedFragment.java:
```diff
@@ -0,0 +1,216 @@
+
+    @Override
+    public void setPresenter(NewsFeedContract.Presenter presenter) {
+        Log.d(TAG, "setPresenter: ");
```
Do we need this logging?

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedFragment.java:
```diff
@@ -0,0 +1,216 @@
+    @Override
+    public void onResume() {
+        super.onResume();
+        updateCategoryNameOnDescription(Category.getDisplayName(Category.ALL));
```
Why is it done by hand? A view shouldn't do anything itself - only render data.

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedFragment.java:
```diff
@@ -0,0 +1,216 @@
+        initRecyclerView();
+        favoriteNewsButton.setOnClickListener(v -> {
+            presenter.invertFavoriteState();
+
```
Yet another unnecessary empty line.

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedFragment.java:
```diff
@@ -0,0 +1,216 @@
+
+        });
+        categoryButton.setOnClickListener(v -> {
+            Navigator.openCategoryDialogFragment(this, Objects.requireNonNull(getActivity()));
```
This should call a presenter which will call a corresponding `View` method to show a dialog.

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedFragment.java:
```diff
@@ -0,0 +1,216 @@
+    @Override
+    public void hideProgressBar() {
+        progressBar.setVisibility(ProgressBar.GONE);
+
```
Do we need this empty line?

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedFragment.java:
```diff
@@ -0,0 +1,216 @@
+
+    @Override
+    public void updateNewsListUI(List<NewsItem> news) {
+        Log.d(TAG, "updateNewsListUI: ");
```
1. I don't think we need any kind of logging here.
2. This logic is a part of the presenter. A presenter should decide what to do and call a corresponding `View` method (and pass a parameter if required).

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedFragment.java:
```diff
@@ -0,0 +1,216 @@
+
+    @Override
+    public void showNewsDetailsUI(String url) {
+        Navigator.openWebView(url, context);
```
This should call a presenter which will use a navigator to process an action.

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedFragment.java:
```diff
@@ -0,0 +1,216 @@
+
+    @Override
+    public void onCategorySelected(Category selectCategory) {
+        presenter.setCategory(selectCategory);
```
This should call a `presenter.onCategorySelected` method which will call required `View` methods and process `selectedCategory`.

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedPresenter.java:
```diff
@@ -0,0 +1,175 @@
+    private MainApplication mainApplication;
+    private NewsStorage newsStorage;
+
+    NewsFeedPresenter(NewsFeedContract.View tasksView, MainApplication mainApplication) {
```
Why do you use default visibility?

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedPresenter.java:
```diff
@@ -0,0 +1,175 @@
+
+    private void initNewsStorage() {
+        NewsItemDAO newsItemDAO = mainApplication.getNewsDataBase().newsItemDAO();
+        newsStorage = new NewsStorage(newsItemDAO);
```
Can't we move this initialization to the storage constructor and just pass a context there?

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedPresenter.java:
```diff
@@ -0,0 +1,175 @@
+    }
+
+    private void initNewsItemListCallback() {
+        // TODO: 17.03.2020 extract in network layer
```
Time to resolve this TODO and move this code somewhere.

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedPresenter.java:
```diff
@@ -0,0 +1,175 @@
+
+    private void performCall() {
+        Log.d(TAG, "performCall: ");
+        NewsApiRepository newsApiRepository = new NewsApiRepository(mainApplication);
```
Too difficult name - we can call this variable a `repository`.

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedPresenter.java:
```diff
@@ -0,0 +1,175 @@
+
+    private List<NewsItem> getNewsItemListByResponse(Response<NewsApiResponse> response, Category category) {
+        List<NewsItem> news = new ArrayList<>();
+        List<NewsApiResponseItem> newsApiResponseItems = Objects.requireNonNull(response.body()).getNewsApiResponseItemList();
```
The same here - we can make things simpler by calling a variable `responseItems`.

arudometkin commented on app/src/main/res/values/strings.xml:
```diff
@@ -12,5 +12,6 @@
     <string name="category_sport">Спорт</string>
     <string name="category_technology">Технологии</string>
     <string name="no_data_available">no data available</string>
+    <string name="tv_category_string_prefix">Категория:</string>
```
Do you remember our discussion on resources? `tv_` prefix is not needed here - we should give this resource a meaningful name instead.

VSemikoz committed:
correct Navigator

VSemikoz committed:
add setter for selectCategoryDialog

VSemikoz committed:
remove todo in contract

VSemikoz committed:
extract presenter initialize in fragment

VSemikoz committed:
remove logs

VSemikoz committed:
initial setup from presenter

VSemikoz committed:
remove empty lines

VSemikoz committed:
extract openCategoryDialogFragment() in presenter

VSemikoz committed:
extract updateNewsListUI logic in presenter

VSemikoz committed:
code refactor

VSemikoz committed:
extract openNewsDetails logic in presenter

VSemikoz committed:
simplify logic of openWebView

VSemikoz committed:
extract newsStorage initialization in constructor

VSemikoz committed:
rename newsApiRepository -> repository

VSemikoz committed:
rename tv_category_string_prefix -> display_category_prefix

VSemikoz committed:
rename newsApiResponseItems -> responseItems

VSemikoz committed:
change directory removed empty line

VSemikoz committed:
add Mappers class and mapResponseToNewsItems method in it

VSemikoz committed:
effects of change dir of ActivityUtils class

VSemikoz committed:
add callback into NewsApiRepository

VSemikoz committed:
extract callback logic into NewsApiRepository

VSemikoz committed:
code refactor

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedActivity.java:
```diff
@@ -0,0 +1,29 @@
+        NewsFeedFragment newsFeedFragment = (NewsFeedFragment) getSupportFragmentManager().findFragmentById(R.id.newsfeed_content_frame);
+        if (newsFeedFragment == null) {
+            /* TODO: 18.03.2020 extract mainApplication in constructor */
+            newsFeedFragment = NewsFeedFragment.newInstance(mainApplication);
```
`MainApplication` isn't needed to create an instance of the `Fragment`.

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedFragment.java:
```diff
@@ -0,0 +1,208 @@
+        Bundle args = new Bundle();
+        NewsFeedFragment fragment = new NewsFeedFragment(mainApplication);
+        fragment.setArguments(args);
+        return fragment;
```
This piece of code does nothing. You can simply remove `newInstance` method.

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedFragment.java:
```diff
@@ -0,0 +1,208 @@
+    private TextView descriptionView;
+
+    private NewsFeedFragment(MainApplication mainApplication) {
+        presenter =  new NewsFeedPresenter(this, mainApplication);
```
Can we create a presenter without passing a `mainApplication` there?

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedFragment.java:
```diff
@@ -0,0 +1,208 @@
+        super.onCreate(savedInstanceState);
+        context = Objects.requireNonNull(getActivity()).getApplicationContext();
+        adapter = new NewsFeedAdapter(context);
+        navigator = new Navigator();
```
A navigator should be used from the presenter.

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedFragment.java:
```diff
@@ -0,0 +1,208 @@
+    @Override
+    public void onResume() {
+        super.onResume();
+        presenter.initStartValues();
```
Does your `View` follow SOLID principles? And this method in particular.

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedFragment.java:
```diff
@@ -0,0 +1,208 @@
+                             @Nullable Bundle savedInstanceState) {
+        View root = inflater.inflate(R.layout.newsfeed_frag, container, false);
+
+        recyclerView = root.findViewById(R.id.rv_news_feed);
```
Can we move an initialization of the `Views` somewhere?

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedPresenter.java:
```diff
@@ -0,0 +1,185 @@
+    }
+
+    private void initNewsItemListCallback() {
+        // TODO: 17.03.2020 extract in network layer
```
Is it fixed?

VSemikoz committed:
delete navigator from fragment

VSemikoz committed:
remove context from MainApplication add insrance

VSemikoz committed:
remove newInstance() from Fragment, add ini MainApplication in Presenter

VSemikoz committed:
change access modifier of Presenter constructor

VSemikoz committed:
remove initStartValues() from Fragment's onResume method

VSemikoz committed:
move init of view in separate functions

VSemikoz committed:
delete setter for RequestListener add reqListener par in getNewsFromApi()

VSemikoz committed:
remove initStartValues from contract

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedActivity.java:
```diff
@@ -0,0 +1,26 @@
+    @Override
+    protected void onCreate(@Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.newsfeed_act);
```
BTW, why have you called your layout file a `newsfeed_act`? The best way to call layouts is: `activity_news_feed`, `fragment_news_feed`, `dialog_news_feed`, `news_feed_item`, etc.

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedFragment.java:
```diff
@@ -0,0 +1,202 @@
+        View root = inflater.inflate(R.layout.newsfeed_frag, container, false);
+        initControls(root);
+        favoriteNewsButton.setOnClickListener(v -> presenter.invertFavoriteState());
+        categoryButton.setOnClickListener(v -> presenter.onCategoryButtonClick());
```
These 2 lines should also be moved to `initControls`.

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedPresenter.java:
```diff
@@ -0,0 +1,169 @@
+public class NewsFeedPresenter implements NewsFeedContract.Presenter {
+    private final NewsFeedContract.View view;
+
+    private String TAG = NewsFeedPresenter.class.getName();
```
This field should be constant and should be at the top of the class.

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedPresenter.java:
```diff
@@ -0,0 +1,169 @@
+    private NewsStorage newsStorage;
+    private NewsApiRepository repository;
+
+    public NewsFeedPresenter(NewsFeedContract.View tasksView) {
```
What is `tasksView`?

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedFragment.java:
```diff
@@ -0,0 +1,202 @@
+    @Override
+    public void updateNewsListUI() {
+        presenter.onUpdateNewsList();
+
```
Empty line?

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedFragment.java:
```diff
@@ -0,0 +1,202 @@
+    }
+
+    @Override
+    public void updateNewsListUI() {
```
This method is doing unnecessary work since we can directly pass a list of items to the view to render it.

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedPresenter.java:
```diff
@@ -0,0 +1,169 @@
+
+    @Override
+    public void onUpdateNewsList() {
+        if (news == null || news.isEmpty()) {
```
This can be done without going through the `View`, I mean you can call this method right from the loading logic but you should definitely rename it.

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedPresenter.java:
```diff
@@ -0,0 +1,169 @@
+        newsStorage = new NewsStorage(mainApplication);
+    }
+
+    private void performCall() {
```
`performCall` is not the best name for this method since it doesn't describe what is going on inside.

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedPresenter.java:
```diff
@@ -0,0 +1,169 @@
+                }
+                List<NewsItem> news = Mappers.mapResponseToNewsItems(response, category);
+                newsStorage.insertUnique(news);
+                Log.d(TAG, "onResponse: ");
```
Do we need such logging here?

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedPresenter.java:
```diff
@@ -0,0 +1,169 @@
+            @Override
+            public void onApiRequestFailure(Throwable t) {
+                view.hideProgressBar();
+                Log.d(TAG, "onFailure " + Objects.requireNonNull(t.getMessage()));
```
Your code will fail in runtime once you'll receive an exception with `null` message here. You shouldn't use checks so much.

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/storage/NewsApiRepository.java:
```diff
@@ -1,20 +1,29 @@
-
 public class NewsApiRepository {
     private MainApplication mainApplication;
+    private Callback<NewsApiResponse> callbackNewsItemList;
```
We can remove it. This shouldn't be cached.

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/storage/NewsApiRepository.java:
```diff
@@ -23,6 +32,27 @@ public void getNewsFromApi(Category category, Callback<NewsApiResponse> callback
+            public void onFailure(@NotNull Call<NewsApiResponse> call, @NotNull Throwable t) {
+                listener.onApiRequestFailure(t);
+            }
+        };
```
We can combine these 2 methods as we don't need to cache a callback.

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/utils/ActivityUtils.java:
```diff
@@ -0,0 +1,20 @@
+        checkNotNull(fragmentManager);
+        checkNotNull(fragment);
+        FragmentTransaction transaction = fragmentManager.beginTransaction();
+        transaction.add(frameId, fragment);
```
You're adding your fragment using `frameId` variable which is not the best name as we need a container. A `containerId` looks like a better name for this variable. What do you think?

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/utils/Mappers.java:
```diff
@@ -0,0 +1,22 @@
+import ru.vssemikoz.newsfeed.models.NewsApiResponseItem;
+import ru.vssemikoz.newsfeed.models.NewsItem;
+
+public class Mappers {
```
Can a class with such a name follow SOLID principles? What are you going to place here?

VSemikoz committed:
extract buttonListeners into initControls()

VSemikoz committed:
rename tasksView -> view

VSemikoz committed:
remove empty line

VSemikoz committed:
rename layabout resources

VSemikoz committed:
convert log variable to const

VSemikoz committed:
rename performCall -> processNewsApiResponse

VSemikoz committed:
remove unnecessary logs

VSemikoz committed:
delete callbackNewsItemList, combine getCallbackNewsItemList and initNewsItemListCallback

VSemikoz committed:
simplify updating NewsList on UI

VSemikoz committed:
delete unnecessary onUpdateNewsList() method

VSemikoz committed:
rename frameId -> containerId

VSemikoz committed:
rename and remove:
 - storage dir to data dir
 - Mappers into NewsItemMapper
 - replace NewsItemMapper into storage/mappers/

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedActivity.java:
```diff
@@ -0,0 +1,25 @@
+import ru.vssemikoz.newsfeed.utils.ActivityUtils;
+
+public class NewsFeedActivity extends AppCompatActivity {
+    private String TAG = NewsFeedActivity.class.getName();
```
This field should be constant.

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedFragment.java:
```diff
@@ -0,0 +1,195 @@
+    }
+
+    @Override
+    public void setEmptyViewOnDisplay() {
```
Let's make things simpler like `showEmptyView` / `showList`. You shouldn't show too many framework details in your method names.

arudometkin commented on app/src/main/java/ru/vssemikoz/newsfeed/newsfeed/NewsFeedPresenter.java:
```diff
@@ -0,0 +1,165 @@
+
+    private void loadNewsFromApi() {
+        view.showProgressBar();
+        processNewsApiResponse();
```
`processNewApiResponse` means that you already have a response to be processed. But what you're doing here actually is requesting news so we can rename this method to something like `requestNewsFromApi`.

arudometkin reviewed (approved):
LG%C

VSemikoz committed:
delete log string

VSemikoz committed:
rename:
 - setEmptyViewOnDisplay -> showEmptyView;
 - setRecyclerViewOnDisplay -> showList

VSemikoz committed:
rename processNewsApiResponse -> requestNewsFromApi

arudometkin reviewed (approved):
LGTM
